upt_macports/upt_macports.py

```diff
@@ -229,32 +229,25 @@ def __init__(self):
         self.logger = logging.getLogger('upt')
 
     name = 'macports'
+    pkg_classes = {
+                'pypi': MacPortsPythonPackage,
+                'cpan': MacPortsPerlPackage,
+                'rubygems': MacPortsRubyPackage,
+                'npm': MacPortsNpmPackage
+            }
 
     def create_package(self, upt_pkg, output=None):
-        pkg_classes = {
-            'pypi': MacPortsPythonPackage,
-            'cpan': MacPortsPerlPackage,
-            'rubygems': MacPortsRubyPackage,
-            'npm': MacPortsNpmPackage
-        }
-
         try:
             self.frontend = upt_pkg.frontend
-            pkg_cls = pkg_classes[upt_pkg.frontend]
+            pkg_cls = self.pkg_classes[upt_pkg.frontend]
         except KeyError:
             raise upt.UnhandledFrontendError(self.name, upt_pkg.frontend)
         packager = pkg_cls()
         packager.create_package(upt_pkg, output)
 
     def package_versions(self, name):
-        pkg_classes = {
-            'pypi': MacPortsPythonPackage,
-            'cpan': MacPortsPerlPackage,
-            'rubygems': MacPortsRubyPackage,
-            'npm': MacPortsNpmPackage
-        }
         try:
-            port_name = pkg_classes[self.frontend](
+            port_name = self.pkg_classes[self.frontend](
             )._normalized_macports_name(name)
```

Comment:
You are instantiating a new upt.Package  object. Why remember self.pkg_classes? Why not keep a handle on "packager" in create_package()?

Reply:
Sorry, but I didn't get you.

Reply:
I am not sure either; @Steap can you educate us? ;)

Anyway, let's resolve this specific issue (#41) right now and revisit the other point (#43,  about where/how to centralize the naming of ports/directories/dependencies) in a later PR. We should really merge this PR soon so that we can move on to the "update" PR - almost there!

Reply:
Yeah agree, also we should address the [version issue](https://github.com/macports/upt-macports/issues/44#issuecomment-513349606) here.

I'll push the changes in few min.

Reply:
I agree that this should be fixed (haven't looked carefully at it though), but if you're description is correct it's not something we should fix here. That sounds like an issue with MacPorts not with the code you have, or...? 

If I am correct about this, then it has nothing to do with this PR, so let's not push changes to this PR.

Reply:
yeah ok.

Reply:
This changes should work perfectly now, the changes I mentioned before that I did when adding update feature is just having `package_versions` take an extra argument as `frontend name` but since upt-core has already merged this part of code so we should not change it now....we can change it while merging update feature instead.

Reply:
When you write:

```python
 port_name = self.pkg_classes[self.frontend]()._normalized_macports_name(name)
```
you are creating a upt.Package() instance that you use only once and then throw away.

You could probably keep the package object by using it as an attribute (self.package = self.pkg_classes[self.frontend]() in create_package()).

Reply:
If I understand this correctly then this will only work if `create_package` is called before `package_versions` or `update_package`(in case of update).

But as you can see in upt-core for update feature we will call `package_versions` before `update_package` and we actually never call `create_package`

Reply:
By the way, I meant "MacPortsPackage", not "upt.Package", sorry :)

Hmm, okay, that's a good point. Maybe we'll change Backend.__init__() to add frontend as an argument, which might be the cleanest way of doing this, but I'm not sure. I'll have to think about this :)

Reply:
So should we get this PR merged now as it's in working state and we change this later on?

Reply:
Yes, that would be fine with me! @Steap can we work on finalizing PR 47 and this one today? That would make reviewing the "update PR" easier since that has several of these commits mixed in..

Reply:
@Korusuke Can you rebase this on top of master, and test it with the latest commits of upt? I released 0.10.1 yesterday.

Reply:
Yep, it works perfectly!